morello-c64-secure/impls.s

```diff
@@ -0,0 +1,238 @@
+        .size A, A_end-A
+
+
+        // B: C64-Secure, with conservative csp bounds restriction.
+        //
+        // Add padding below the frame record to ensure that csp does not cover
+        // it. This might waste substantial amounts of stack memory, especially
+        // near the start of the stack.
+	.global B
+	.type B, %function
+B:
+	// Frame record:
+        //    cfp + 32:  CLR / previous PCC
+        //    cfp + 16:  Previous CSP
+        //    cfp +  0:  Previous CFP
+        mov c10, csp
+	str cfp, [csp, #-48]!
+	stp c10, clr, [csp, #16]
+        // Point cfp to the frame record, and restrict its bounds.
+        scbnds cfp, csp, #48
+
+        // The requested length is gcvalue(cfp) - gcbase(csp).
+        // We'll round that down to get the representable length.
+        scoff c10, csp, xzr
+        sub x11, fp, x10
+        // rrlen rounds the length _up_, so use rrmask instead and round the
+        // length down manually.
+        rrmask x12, x11
+        and x12, x11, x12
+        // We assume that since the original stack must have been larger than
+        // this, its base is suitably aligned.
+        // TODO: Prove that this is sound.
+        scbndse c10, c10, x12
+        sub x12, x12, w0, uxtw #4   // Allocate locals.
+        scoff csp, c10, x12
+
+1:      mov c0, cfp
+        mov c1, csp
+        adr c2, 2f
+        bl what_next
+        // w0: 'locals' for the callee.
+        // c1: The callee's address (if any).
+2:      chktgd c1
+        b.cc 3f
+        blrs c1
+        b 1b  // Keep going until it's time to return.
+3:
+        // Epilogue: Unravel everything.
+        ldp c10, clr, [cfp, #16]
+        ldr cfp, [cfp]
+        mov csp, c10
+	ret clr
+B_end:
+        .size B, B_end-B
+
+
+        // C: C64-Secure v2, with precisely-restricted csp (and full cfp).
+        //
+        // Give CFP full-stack bounds, and derive a smaller, exact CSP for each
+        // function.
+        //
+        // This assumes that, on entry, CFP's bounds extend to the stack limit.
+        .global C
+        .type C, %function
+C:
+        // Frame record:
+        //    cfp + 16: CLR / previous PCC
+        //    cfp +  0: Previous CFP with previous CSP's bounds
+        mov c16, csp
+        scvalue c16, c16, fp
+        stp c16, clr, [csp, #-32]!
+        // Move the value of csp into cfp, but keep cfp's bounds.
+        mov x16, sp
+        scvalue cfp, cfp, x16
+
+        // Derive a new csp from the new cfp.
+        // TODO: For now, we just allow 16KB per function, but we could pick the
+        // longest length that guarantees 16-byte alignment (E=4).
+        mov x16, #(16 * 1024)
+        sub x17, sp, x16
+        scvalue c17, cfp, x17
+        scbndse c17, c17, x16
+        sub x16, sp, w0, uxtw #4    // Allocate locals.
+        scvalue csp, c17, x16
+        // TODO: The very last section of stack will be unusable because
+        // scbnds{e} fails if the required bounds exceed the bounds of <Cn>.
```

Comment:
Not before merge, since what is implemented is sufficient for the demo, but we might decide to address them in some future update if the corner-cases are interesting.

I normally use "TODO" for such things, but if the project prefers a different marker (like `NB`), that's fine too.

Reply:
I think `NB` would be clearer! Like you say, what's implemented is fine for the demo itself.